fuse/ipns/ipns_unix.go

```diff
@@ -309,279 +271,130 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
-	return nil, nil
+func (fi *File) Release(ctx context.Context, req *fuse.ReleaseRequest) error {
+	return fi.fi.Close()
 }
 
-func (n *Node) Create(ctx context.Context, req *fuse.CreateRequest, resp *fuse.CreateResponse) (fs.Node, fs.Handle, error) {
-	if IpnsReadonly {
-		log.Debug("Attempted to call Create on a readonly filesystem.")
-		return nil, nil, fuse.EPERM
-	}
-
+func (dir *Directory) Create(ctx context.Context, req *fuse.CreateRequest, resp *fuse.CreateResponse) (fs.Node, fs.Handle, error) {
 	// New 'empty' file
-	nd := &mdag.Node{Data: ft.FilePBData(nil, 0)}
-	child := n.makeChild(req.Name, nd)
-
-	nnode := n.Nd.Copy()
+	nd := &dag.Node{Data: ft.FilePBData(nil, 0)}
+	err := dir.dir.AddChild(req.Name, nd)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	err := nnode.AddNodeLink(req.Name, nd)
+	child, err := dir.dir.Child(req.Name)
 	if err != nil {
 		return nil, nil, err
 	}
-	if n.parent != nil {
-		err := n.parent.update(n.name, nnode)
-		if err != nil {
-			log.Criticalf("Error updating node: %s", err)
-			// Can we panic, please?
-			return nil, nil, err
-		}
+
+	fi, ok := child.(*nsfs.File)
+	if !ok {
+		return nil, nil, errors.New("child creation failed")
 	}
-	n.Nd = nnode
-	n.wasChanged()
 
-	return child, child, nil
+	nodechild := &File{fi: fi}
+	return nodechild, nodechild, nil
 }
 
-func (n *Node) Remove(ctx context.Context, req *fuse.RemoveRequest) error {
-	if IpnsReadonly {
-		return fuse.EPERM
-	}
-
-	nnode := n.Nd.Copy()
-	err := nnode.RemoveNodeLink(req.Name)
+func (dir *Directory) Remove(ctx context.Context, req *fuse.RemoveRequest) error {
+	err := dir.dir.Unlink(req.Name)
 	if err != nil {
 		return fuse.ENOENT
 	}
-
-	if n.parent != nil {
-		err := n.parent.update(n.name, nnode)
-		if err != nil {
-			log.Criticalf("Error updating node: %s", err)
-			return err
-		}
-	}
-	n.Nd = nnode
-	n.wasChanged()
 	return nil
 }
 
-func (n *Node) Rename(ctx context.Context, req *fuse.RenameRequest, newDir fs.Node) error {
-	if IpnsReadonly {
-		log.Debug("Attempted to call Rename on a readonly filesystem.")
-		return fuse.EPERM
+// Rename implements NodeRenamer
+func (dir *Directory) Rename(ctx context.Context, req *fuse.RenameRequest, newDir fs.Node) error {
+	cur, err := dir.dir.Child(req.OldName)
+	if err != nil {
+		return err
 	}
 
-	var mdn *mdag.Node
-	for _, l := range n.Nd.Links {
-		if l.Name == req.OldName {
-			mdn = l.Node
-		}
-	}
-	if mdn == nil {
-		log.Critical("nil Link found on rename!")
-		return fuse.ENOENT
+	err = dir.dir.Unlink(req.OldName)
+	if err != nil {
+		return err
 	}
-	n.Nd.RemoveNodeLink(req.OldName)
 
 	switch newDir := newDir.(type) {
-	case *Node:
-		err := newDir.Nd.AddNodeLink(req.NewName, mdn)
+	case *Directory:
+		nd, err := cur.GetNode()
 		if err != nil {
 			return err
 		}
+
+		err = newDir.dir.AddChild(req.NewName, nd)
+		if err != nil {
+			return err
+		}
+	case *File:
+		log.Critical("Cannot move node into a file!")
```

Comment:
no, im serious, you cant actually cause this to happen. i would have to do `mv file1 file2` which would just overwrite file2 with file1, not try to move file1 under file2